File: 1719-number-of-ways-to-reconstruct-a-tree/1719-number-of-ways-to-reconstruct-a-tree.py
from collections import defaultdict

class Solution:
    def checkWays(self, P):
        adj = defaultdict(set)
        for u, v in P:
            adj[u].add(v)
            adj[v].add(u)

        def explore_component(node, component, visited):
            component.add(node)
            visited.add(node)
            for neighbor in adj[node]:
                if neighbor not in visited:
                    explore_component(neighbor, component, visited)

        def analyze(nodes):
            degree_groups = defaultdict(list)
            max_degree = len(nodes) - 1
            for node in nodes:
                degree_groups[len(adj[node])].append(node)

            if len(degree_groups[max_degree]) == 0:
                return 0
            root = degree_groups[max_degree][0]

            for neighbor in adj[root]:
                adj[neighbor].remove(root)

            connected_components = defaultdict(set)
            visited_nodes = set()
            group_counter = 0
            for node in nodes:
                if node != root and node not in visited_nodes:
                    explore_component(node, connected_components[group_counter], visited_nodes)
                    group_counter += 1

            results = [analyze(connected_components[i]) for i in connected_components]
            if 0 in results:
                return 0
            if 2 in results:
                return 2
            if len(degree_groups[max_degree]) > 1:
                return 2
            return 1

        return analyze(set(adj.keys()))

File: 1719-number-of-ways-to-reconstruct-a-tree/1719-number-of-ways-to-reconstruct-a-tree.java
class Solution {
    public int checkWays(int[][] pairs) {
        System.out.println(pairs.length);
        HashMap<Integer, Set<Integer>> map = new HashMap<>();
        HashMap<Integer, Integer> prt = new HashMap<>();
        sum = 0;
        
        for(int[] pair:pairs){
            if(!map.containsKey(pair[0])){
                map.put(pair[0], new HashSet<>());
                prt.put(pair[0], pair[0]);
            }
            if(!map.containsKey(pair[1])){
                map.put(pair[1], new HashSet<>());
                prt.put(pair[1], pair[1]);
            }
            map.get(pair[0]).add(pair[1]);
            map.get(pair[1]).add(pair[0]);
        }

        HashMap<Integer, ArrayList<Integer>> adj = new HashMap<>();
        int root = -1;
        while(!map.isEmpty()){
            int ele = -1;
            int sz = -1;
            for(int ky: map.keySet()){
                if(sz < map.get(ky).size()){
                    sz = map.get(ky).size();
                    ele = ky;
                }
            }
            if(root == -1) {
                root = ele;
                if(sz != prt.size()-1) return 0;
            }
            
            if(!adj.containsKey(ele)) adj.put(ele, new ArrayList<>());
            if(ele != prt.get(ele)){
                adj.get(prt.get(ele)).add(ele);
            }

            for(int num: map.get(ele)){
                prt.replace(num, ele);
                map.get(num).remove(ele);
            }

            map.remove(ele);
        }

        validity(adj, root, 0);
        if(sum != pairs.length) return 0;
        
        for(ArrayList<Integer> ele: adj.values()){
            if(ele.size() == 1) return 2;
        }
        
        return 1;
    }
    
    int sum;
    public void validity(HashMap<Integer, ArrayList<Integer>> map, int node, int cnt){
        for(int nxt: map.get(node)){
            validity(map, nxt, cnt+1);
        }
        
        sum += cnt;
    }
}